File: compositions/extdemo/bin/devices.rb
#!/usr/bin/ruby

require 'net/http'
require 'json'

uri = URI.parse("http://localhost:2000/extdemo")

devices = {
    'BEACON 814' => 'beacon',
    'Office' => 'home',
    'Laptop' => 'laptop',
    'Inbox' => 'message',
    'Bluetooth' => 'mic',
    'iPhone' => 'phone',
    'Charging' => 'power',
    'Brother 429' => 'printer',
    'Kitchen Lamp' => 'smartbulb',
    'Admin' => 'terminal',
    'LED H4500' => 'tv',
    'Wallet' => 'wallet',

    'WiFi' => 'beacon',
    'Home Network' => 'home',
    'Desktop' => 'laptop',
    'SMS' => 'message',
    'Android' => 'phone',
    'Air Print' => 'printer',
    'Lights' => 'smartbulb',
    'Smart TV' => 'tv'
}

loop do

  selected = devices.to_a.sample
  action = %w(join join part).sample

  req = Net::HTTP::Post.new(uri, initheader = {'Content-Type' =>'application/json'})

  body = {
      "rpcid" => "devbus",
      "type" => "method",
      "method" => "notify",
      "args" => [selected.first, selected.last, action]
  }.to_json
  puts body

  req.body = body


  res = Net::HTTP.start(uri.hostname, uri.port) do |http|
    http.request(req)
  end

  puts res
  sleep(3)

end

File: compositions/extdemo/devices.js
// Access the `active` attribute on the command line with these:
//
// curl -d '{"rpcid":"devbus", "type":"attribute", "attribute":"active", "get":true}' http://localhost:2000/extdemo
//
// curl -d '{"rpcid":"devbus", "type":"attribute", "attribute":"active", "value":{ "My Phone":"phone", "SmartLamp":"smartbulb", "Brother 101":"printer" }}' http://localhost:2000/extdemo
//
// Use `notify` method via the command line with one of:
//
// curl -d '{"rpcid":"devbus", "type":"method", "method":"notify", "args":["ID", "TYPE", "join"]}' http://localhost:2000/extdemo
//
// curl -d '{"rpcid":"devbus", "type":"method", "method":"notify", "args":["ID", "TYPE", "part"]}' http://localhost:2000/extdemo

define.class(function (server) {

    this.attributes = {active: {type: Object, value:{}}};

    this.notify = function(id, type, action) {
        var active = this.active;

        if (action == 'join') {
            active[id] = type;
        } else if (action == 'part') {
            delete active[id];
        }

        this.active = active;
    }

});
